Change Ruby Logging

titusfortner opened the pull request:
### Description

Change default logging level to `:info` and provide more granular filtering options

### Motivation and Context

I'm looking at how logging is being done in other languages, and as part of putting this documentation together - https://www.selenium.dev/documentation/webdriver/troubleshooting/logging/ I thought we should change our implementation to match the other languages.

### Details:

To better manage the flood of details now in `:debug` mode, I've added `#select` and `#reject` methods where you can opt in or opt out of specific types of logs.

```rb
Selenium::WebDriver.logger.level = :debug
Selenium::WebDriver.logger.select(:selenium_manager)
```

```rb
Selenium::WebDriver.logger.reject(%i[deprecations logger_info])
```

Let me know if all of these choices make sense.

p0deje commented:
I'm looking through https://www.selenium.dev/documentation/webdriver/troubleshooting/logging/, but I don't see what exactly we should match in Ruby. Can you point to Java/Python/etc examples?

titusfortner commented:
I meant the levels.

* Actionable items
* Useful information
* Debugging details

Right now `:warn` is a mix of info and actionable
And `:info` is essentially another level of `:debug` details

I wish Ruby had more "debug" levels like the other languages, but alas.

p0deje commented on rb/lib/selenium/webdriver/common/driver_finder.rb:
```diff
@@ -27,8 +27,7 @@ def self.path(options, klass)
         path ||= begin
           SeleniumManager.driver_path(options)
-        rescue StandardError => e
-          WebDriver.logger.warn("Unable obtain driver using Selenium Manager\n #{e.message}")
```
Why do we remove it?

titusfortner replied:
The idea was we don't have to log this if we're raising the error message, but I don't think we have the right error message yet, so better to keep this for now.

p0deje commented on rb/lib/selenium/webdriver/common/logger.rb:
```diff
@@ -83,32 +91,59 @@ def io
       #
-      def ignore(id)
-        Array(id).each { |ignore| @ignored << ignore }
+      def reject(id)
+        Array(id).each { |key| @reject << key }
+      end
+      alias ignore reject
```
What do you think about using splat arguments for those?

```suggestion
      def reject(*ids)
        @reject += ids
      end
      alias ignore reject
```

titusfortner replied:
Hmm, would be nice to support `ignore(:foo, :bar, :baz)` as well as `ignore(%i[foo bar baz])`

titusfortner replied:
but I guess if choosing one, arguments are better than Array?

p0deje replied:
I think it's the perfect case for splat arguments.

titusfortner replied:
```
        @ignored += Array(ids).flatten
```

p0deje commented on rb/lib/selenium/webdriver/common/logger.rb:
```diff
@@ -83,32 +91,59 @@ def io
+      #
+      # @param [Array, Symbol] id
+      #
+      def select(id)
```
Ditto.

p0deje commented on rb/lib/selenium/webdriver/common/logger.rb:
```diff
@@ -83,32 +91,59 @@ def io
       #
-      def ignore(id)
-        Array(id).each { |ignore| @ignored << ignore }
+      def reject(id)
```
Frankly, having `select/reject` methods look strange to me on logger because those are typically methods on enumerable. I can see why they are there, but I would need to consult the documentation to understand their purpose. What do you think about using more discoverable names such as `allow/ignore`?

titusfortner replied:
yeah, I wasn't sure about these names, and I think they are confusing since they are different from how they are used elsewhere. I couldn't think of a good "opposite" for `ignore`, but `allow` makes sense.

p0deje commented on rb/lib/selenium/webdriver/common/logger.rb:
```diff
@@ -133,8 +169,9 @@ def deprecate(old, new = nil, id: [], reference: '', &block)
+        message += " #{yield}" if block_given?
 
-        warn message, &block
+        @logger.warn { message }
```
The code is very similar to `#discard_or_log`. I suggest you refactor it so that it accepts `@reject/@select` as the parameters. Then you can simply call this method with an extra `:deprecations` id.

p0deje commented on rb/lib/selenium/webdriver/common/logger.rb:
```diff
@@ -38,20 +38,28 @@ class Logger
+      def initialize(progname = 'Selenium', ignored: nil, select: nil, reject: nil)
         @logger = create_logger(progname)
-        @ignored = Array(ignored)
+        @reject = Array(Array(reject) + Array(ignored))
```
I suggest you use sets for `@reject/@select`. This way you would have cleaner APIs (https://ruby-doc.org/stdlib-2.7.1/libdoc/set/rdoc/Set.html) when checking ids and won't have to worry about uniqueness and order.

titusfortner replied:
I guess I would prefer to use Arrays because there isn't exactly a "problem" with duplication, and Ruby makes arrays so much easier/flexible with splats, etc.

titusfortner commented:
I fixed the things you suggested, including changing the names. I'd prefer Arrays to Sets still just because that's what people are more used to... Methods now accept one or more params and Arrays. 

I also added logic for setting the default level in constructor. Webdrivers gem and Watir both use the Selenium Logger, and don't want to force the change on them?

p0deje commented:
> I fixed the things you suggested, including changing the names. I'd prefer Arrays to Sets still just because that's what people are more used to... Methods now accept one or more params and Arrays.

My proposal was to convert arrays to sets internally, not to force people into using sets.

p0deje reviewed (approved):
LGTM but CI is failing.
